go/disk.go

```diff
@@ -0,0 +1,352 @@
+	// Path specifies where the image file will be.
+	Path string `json:"path"`
+	// Size specifies the size of the disk image.  Used if the image needs to be created.
+	Size int `json:"size"`
+	// Format is passed as-is to the driver.
+	Format string `json:"format"`
+	// Trim specifies whether we should trim the image file.
+	Trim bool `json:"trim"`
+	// QcowToolPath is the path to the binary to use to manage this image.
+	// Defaults to "qcow-tool" when empty.
+	QcowToolPath   string
+	OnFlush        string
+	CompactAfter   int
+	KeepErased     int
+	RuntimeAsserts bool
+	Stats          string
+}
+
+// GetPath returns the location of the disk image file.
+func (d *QcowDisk) GetPath() string {
+	return d.Path
+}
+
+// SetPath changes the location of the disk image file.
+func (d *QcowDisk) SetPath(p string) {
+	d.Path = p
+}
+
+// GetSize returns the desired size of the disk image file.
+func (d *QcowDisk) GetSize() int {
+	return d.Size
+}
+
+// String returns the path.
+func (d *QcowDisk) String() string {
+	return d.Path
+}
+
+// QcowTool prepares a call to qcow-tool on this image.
+func (d *QcowDisk) QcowTool(verb string, args ...string) *exec.Cmd {
+	path := d.QcowToolPath
+	if path == "" {
+		path = "qcow-tool"
+	}
+	return exec.Command(path, append([]string{verb, d.Path}, args...)...)
+}
+
+// Exists if the image file exists.
+func (d *QcowDisk) Exists() bool {
+	return exists(d)
+}
+
+// Ensure creates the disk image if needed, and resizes it if needed.
+func (d *QcowDisk) Ensure() error {
+	return ensure(d)
+}
+
+// Create a disk with the given size in MiB
+func (d *QcowDisk) create() error {
+	cmd := d.QcowTool("create", "--size", fmt.Sprintf("%dMiB", d.Size))
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	return cmd.Wait()
+}
+
+// Query the current virtual size of the disk in MiB
+func (d *QcowDisk) getFileSize() (int, error) {
+	// Return a failure if the file doesn't exist yet
+	if _, err := os.Stat(d.Path); err != nil {
+		return 0, err
+	}
+	cmd := d.QcowTool("info", "--filter", ".size")
+	out, err := cmd.Output()
+	size, err := strconv.ParseInt(strings.TrimSpace(string(out)), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+	return int(size / mib), nil
+}
+
+// Resize the virtual size of the disk
+func (d *QcowDisk) resize() error {
+	cmd := d.QcowTool("resize", "--size", fmt.Sprintf("%dMiB", d.Size))
+	return cmd.Run()
+}
+
+func (d *QcowDisk) sizeString() string {
+	s, err := d.getFileSize()
+	if err != nil {
+		return fmt.Sprintf("cannot get size: %v", err)
+	}
+	return fmt.Sprintf("%vMiB", s)
+}
+
+// compact the disk to shrink the physical size.
+func (d *QcowDisk) compact(lockFh *os.File) error {
+	log.Infof("Compact: %q... (%v)", d, d.sizeString())
+	cmd := d.QcowTool("compact")
+	// Pass a reference to the file lock to the compact program
+	cmd.ExtraFiles = []*os.File{lockFh}
+	if err := cmd.Run(); err != nil {
+		if err.(*exec.ExitError) != nil {
+			return errors.New("Failed to compact qcow2")
```

Comment:
`d.QcowTool` just calls `exec.Command` which I don't think does anything with `cmd.Stdout` (or `Stderr`) by default, meaning they get redirected to `/dev/null`. I think we probably want the error messages from the tool to go somewhere.

Looking at the callers of `QcowTool` I think they might all be happy with simply directing the output to Stdio, i.e. `QcowTool` could set `Stdout` and `Stderr` to `os.Stdout` and `os.Stderr` respectively. If you want to put the output somewhere other than stdio (to the logs perhaps) then you would want to use `cmd.CombinedOutput` (I think that is better that `cmd.Output` for the cases here) and log the output in the error case (I think in all the users not just this one).

Reply:
Sounds great!  Will do that.